src/views/withNavigationFocus.js

```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import hoistStatics from 'hoist-non-react-statics';
+
+export default function withNavigationFocus(Component) {
+  class ComponentWithNavigationFocus extends React.Component {
+    static displayName = `withNavigationFocus(${Component.displayName ||
+      Component.name})`;
+
+    state = {
+      isFocused: false,
+    };
+
+    componentDidMount() {
+      if (this.props.navigation) {
+        this.subs = [
+          this.props.navigation.addListener('didFocus', () =>
+            this.setState({ isFocused: true })
+          ),
+          this.props.navigation.addListener('willBlur', () =>
+            this.setState({ isFocused: false })
+          ),
+        ];
+      } else {
+        console.warn(
+          'withNavigationFocus wrapped component did not receive navigation from props, so isFocused prop will remain false'
+        );
+      }
+    }
+
+    componentWillUnmount() {
+      this.subs && this.subs.forEach(sub => sub.remove());
+    }
+
+    render() {
+      return (
+        <Component
+          {...this.props}
+          isFocused={this.state.isFocused}
+          ref={this.props.onRef}
```

Comment:
This will not work for function components. We need to check if the component is a function component or not before attaching the ref, and only attach the ref to class components.

On another note, I have seen a `getWrappedInstance` method to be a more common way to expose the wrapped component's ref, so maybe better to use it instead of `this.props.onRef` for better ecosystem compatibility.

Reply:
Here is an example HOC where I do it:
https://github.com/callstack/react-native-paper/blob/master/src/core/withTheme.js#L84
https://github.com/callstack/react-native-paper/blob/master/src/core/withTheme.js#L114

Reply:
We already used onRef somewhere else cc @brentvatne (although I don't feel strongly either way)

Reply:
cc @brentvatne

Reply:
I prefer `getWrappedInstance` compared to `this.props.onRef` as it is used in react-intl and react-redux (but it requires a `{withRef: true}` config on the HOC. I used onRef because it's already what is available on `withNavigation` so didn't want to use something different and keep the API consistent.


Also worth noting that I don't like much the api of react-intl as it does not play well with `compose` because the config is 2nd arg. I prefer allowing `withNavigationFocus(config)(Comp)` and withNavigationFocus(Comp)` instead of withNavigationFocus(Comp,config)`. What do you think about it?

Reply:
> it requires a {withRef: true} config on the HOC

Yeah, feels kinda unnecessary, but no idea why it's that way.

> I prefer allowing withNavigationFocus(config)(Comp)

+1, but what config?

Reply:
Do we need really need a config? Isn't it enough to detect if a component is function component or class? Or there are other issues with attaching ref which I'm missing?

Reply:
`onRef` is used here https://github.com/react-navigation/react-navigation/blob/67f98b69eb141d9adbc579fb5256914fd5b66ccc/src/views/withNavigation.js#L20